50.IESProfileTest/compute/cdc.comp

```diff
@@ -0,0 +1,189 @@
+	}
+
+	return normalize(position);
+}
+
+vec2 sphericalDirToAngles(vec3 direction)
+{
+	double theta = acos(clamp(direction.z, -1.0, 1.0));
+	double phi = atan(direction.y, direction.x);
+	
+	if(phi < 0)
+		phi += 2 * M_PI;
+	
+	return vec2(theta, phi);
+}
+
+ivec2 implGetVUpperBoundIdx(double angle)
+{
+	uint anglesLength = vAngles.length();
+    int idx = -1;
+	
+    for (int i = 0; i < anglesLength; ++i) {
+        if (vAngles[i] > angle) {
+            idx = i;
+            break;
+        }
+    }
+	
+    return ivec2(idx, anglesLength);
+}
+
+ivec2 implGetHUpperBoundIdx(double angle)
+{
+	uint anglesLength = hAngles.length();
+    int idx = -1;
+	
+    for (int i = 0; i < anglesLength; ++i) {
+        if (hAngles[i] > angle) {
+            idx = i;
+            break;
+        }
+    }
+	
+    return ivec2(idx, anglesLength);
+}
+
+int getVLowerBoundIdx(double angle) 
+{
+    return max(implGetVUpperBoundIdx(angle).x - 1, 0);
+}
+
+int getHLowerBoundIdx(double angle) 
+{
+    return max(implGetHUpperBoundIdx(angle).x - 1, 0);
+}
+
+int getVUpperBoundIdx(double angle)
+{
+	ivec2 returnValue = implGetVUpperBoundIdx(angle);
+    return min(returnValue.x, returnValue.y - 1);
+}
+
+int getHUpperBoundIdx(double angle)
+{
+	ivec2 returnValue = implGetHUpperBoundIdx(angle);
+    return min(returnValue.x, returnValue.y - 1);
+}
+
+double getValue(int i, int j)
+{
+	return data[vAngles.length() * i + j];
+}
+
+double sampleI(double vAngle, double hAngle)
+{
+	double vABack = vAngles[vAngles.length() - 1];
+	double hABack = hAngles[hAngles.length() - 1];
+
+	if (vAngle > vABack)
+		return 0.0;
+	
+	hAngle = hABack == 0.0 ? 0.0 : fmod(hAngle, hABack);
+
+	// bilinear interpolation
+	int j0 = getVLowerBoundIdx(vAngle);
+	int j1 = getVUpperBoundIdx(vAngle);
+	int i0 = getHLowerBoundIdx(hAngle);
+	int i1 = getHUpperBoundIdx(hAngle);
+	
+	double uReciprocal = i1 == i0 ? 1.0 : 1.0 / (hAngles[i1] - hAngles[i0]);
+	double vReciprocal = j1 == j0 ? 1.0 : 1.0 / (vAngles[j1] - vAngles[j0]);
+	
+	double u = (hAngle - hAngles[i0]) * uReciprocal;
+	double v = (vAngle - vAngles[j0]) * vReciprocal;
+	
+	double s0 = getValue(i0, j0) * (1.0 - v) + getValue(i0, j1) * (v);
+	double s1 = getValue(i1, j0) * (1.0 - v) + getValue(i1, j1) * (v);
+	
+	return s0 * (1.0 - u) + s1 * u;
+}
+
+void main()
+{
+	const double VERTICAL_INVERSE = 1.0 / TEXTURE_SIZE;
+	const double HORIZONTAL_INVERSE = 1.0 / TEXTURE_SIZE;
+
+	const ivec2 pixelCoordinates = ivec2(gl_GlobalInvocationID.xy);
+	const ivec2 destinationSize = imageSize(outIESCandelaImage);
+	
+	if (all(lessThan(pixelCoordinates, destinationSize)))
+	{
+		const vec2 uv = vec2(pixelCoordinates.x * VERTICAL_INVERSE, pixelCoordinates.y * HORIZONTAL_INVERSE);
+		vec3 direction = octahedronUVToDir(uv);
+		vec2 sphericalCoordinates = sphericalDirToAngles(direction); // third radius spherical compoment is normalized and skipped
```

Comment:
check two things:
1. if `octahedronUVToDir` gives you normalized vectors as output
2. `sphericalCoordinates` return in domain `-PI/2,PI/2` x `-PI,PI`

Reply:
![image](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/assets/34793522/c9adb3ef-eaf8-4a8f-a9aa-03e4049ad9e0)
![image](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/assets/34793522/c543060b-afa2-437b-87cc-485b205abc3b)
![image](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/assets/34793522/a1a32cda-9328-4408-9bb8-6a0e6339a16d)
it appears we don't pass both tests https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/commit/6a3ca12c9139387a104a0b18f95ff2cba62db4af

Reply:
> > sphericalCoordinates return in domain -PI/2,PI/2 x -PI,PI
> 
> I would check range [0, PI] x [0, 2PI] though, this convention also explains why we do 2PI `phi` shift in `sphericalDirToRadians` if `phi < 0` and naturally maps to IES angles ![image](https://private-user-images.githubusercontent.com/34793522/318265374-1cad8354-3283-4fb8-a9cd-65a6ecf994ec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE4ODc0OTEsIm5iZiI6MTcxMTg4NzE5MSwicGF0aCI6Ii8zNDc5MzUyMi8zMTgyNjUzNzQtMWNhZDgzNTQtMzI4My00ZmI4LWE5Y2QtNjVhNmVjZjk5NGVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzMxVDEyMTMxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhNWFhMzVkMDgzYWEzMDBjYWViMjg4NGQyOGExMGZhNTMyNWZiNjA5MjJhZGYxOTQ1MDM5NWU5NmM5NjQ4NDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.j2S200ZRGdE5S-BaCJ1RhTySCHAz0PzQs5mu6E8eKME)

and now I pass domain tests
![image](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/assets/34793522/f9a7da93-a9d2-4e11-91ff-2dc6ba29c72c)

Reply:
> > sphericalCoordinates return in domain -PI/2,PI/2 x -PI,PI
> 
> I would check range [0, PI] x [0, 2PI] though, this convention also explains why we do 2PI `phi` shift in `sphericalDirToRadians` if `phi < 0` and naturally maps to IES angles ![image](https://private-user-images.githubusercontent.com/34793522/318265374-1cad8354-3283-4fb8-a9cd-65a6ecf994ec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwNDgxNzksIm5iZiI6MTcxMjA0Nzg3OSwicGF0aCI6Ii8zNDc5MzUyMi8zMTgyNjUzNzQtMWNhZDgzNTQtMzI4My00ZmI4LWE5Y2QtNjVhNmVjZjk5NGVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAyVDA4NTExOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyMDhlYjMwMjVmNTEzN2EzM2NjMThmZGM3ZmI4MjUxOTI1ODhkMWZlODI5NTNkYTkyN2FjNzM4OTIwY2EyZGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hFAZcsE7Co5QrTAeZasDCJZvBDMtns_X3YeLsa8c6O0)

**nope**, I'm withdrawing from this - `[-PI,PI]` is much better for mirror & mirror repeat wrap

however a small correction here - theta is `[0, PI]` not `[-PI/2,PI/2]`

Reply:
resolved

Reply:
How does a -PI/2,PI/2 theta make sense?

Theta is a polar angle, measured from a pole, not equator, this then turns your z coordinate into a sine instead of a cosine everyone expects

Reply:
> How does a -PI/2,PI/2 theta make sense?
> 
> Theta is a polar angle, measured from a pole, not equator, this then turns your z coordinate into a sine instead of a cosine everyone expects

$$
\theta \in [0, \pi], \phi \in [-\pi,\pi]
$$

[here](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/104/files#diff-7434f97ac7bdb6f4d605ba820188cc3c84a3b259e70f933c9b165972ca877e6cR71) & [here](https://github.com/Devsh-Graphics-Programming/Nabla/pull/671/files#diff-80257c9cee7050d64220c01448e9b1e311778e3ead3388fd2a4c64813caa4638R87)